### PR TITLE
python312Packages.python-hglib: Remove nose dependency; modernize

### PR DESCRIPTION
**Pull request by pyrox0**
## Description of changes
Part of #326513 

This required a few more patches and fixes to other packages than I would like, but this does allow all dependents of python-hglib to build and test properly.

Updates: 
- python312Packages.glean-parser: 4.1.2 -> 4.3.0
- python312Packages.glean-sdk: 52.7.0 -> 60.4.0
Changes:
- `mozphab` has 3 failing tests disabled
- python312Packages.python-hglib is now fetched directly from the upstream repo and has patches applied to remove nose and fix failing tests.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/glean-parser/default.nix [hunk 1]
```diff
@@ -8,43 +8,39 @@
...
   src = fetchPypi {
     pname = "glean_parser";
     inherit version;
-    hash = "sha256-OL59Tg+rD4M0DjQnkU4IqGMcf6sIioxg6bVDyrbqgww=";
+    hash = "sha256-tI1kMCn7gksLdq2ytKAOiKSd5OxHmsnFrdUsUR6b5IE=";
   };
 
   postPatch = ''
     substituteInPlace setup.py \
       --replace "pytest-runner" "" \
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/glean-parser/default.nix:29]:
```suggestion
      --replace "pytest-runner" ""
```

File: pkgs/development/python-modules/glean-parser/default.nix [hunk 2]
```diff
@@ -8,43 +8,39 @@
...
-    hash = "sha256-OL59Tg+rD4M0DjQnkU4IqGMcf6sIioxg6bVDyrbqgww=";
+    hash = "sha256-tI1kMCn7gksLdq2ytKAOiKSd5OxHmsnFrdUsUR6b5IE=";
   };
 
   postPatch = ''
     substituteInPlace setup.py \
-      --replace "pytest-runner" "" \
-      --replace "MarkupSafe>=1.1.1,<=2.0.1" "MarkupSafe>=1.1.1"
+      --replace-fail "pytest-runner" ""
   '';
```
Comment (emilazy) [inline pkgs/development/python-modules/glean-parser/default.nix:30]: Can you use `pythonRemoveDeps` for this?

File: pkgs/development/python-modules/python-hglib/default.nix [hunk 3]
```diff
@@ -1,24 +1,39 @@
...
+      excludes = [ "heptapod-ci.yml" ];
+      url = "https://repo.mercurial-scm.org/python-hglib/raw-rev/8341f2494b3f";
+      hash = "sha256-4gicVCAH94itxHY0l8ek0L/RVhUrw2lMbbnENbWrV6U=";
+    })
+    (fetchpatch2 {
+      name = "fix-tests.patch";
+      url = "https://repo.mercurial-scm.org/python-hglib/raw-rev/a2afbf236ca8";
+      hash = "sha256-T/yKJ8cMMOBVk24SXwyPOoD321S1fZEIunaPJAxI0KI=";
+    })
+  ];
```
Comment (emilazy) [inline pkgs/development/python-modules/python-hglib/default.nix:32]: Could you use the full commit hashes here for URL stability?

**Review by emilazy — CHANGES_REQUESTED** [review]
Great work, thank you! Just a couple nits.

File: pkgs/development/python-modules/glean-parser/default.nix (hunk 2 above)
Comment (pyrox0) [reply pkgs/development/python-modules/glean-parser/default.nix:30]: I've tried using pythonRemoveDeps, and I always get a missing dependency error for the package.

**Review by emilazy — APPROVED** [review]
Result of `nixpkgs-review pr 330739` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>21 packages built:</summary>
  <ul>
    <li>mozillavpn</li>
    <li>mozphab</li>
    <li>mozphab.dist</li>
    <li>powerline (python312Packages.powerline)</li>
    <li>powerline.dist (python312Packages.powerline.dist)</li>
    <li>python311Packages.glean-parser</li>
    <li>python311Packages.glean-parser.dist</li>
    <li>python311Packages.glean-sdk</li>
    <li>python311Packages.glean-sdk.dist</li>
    <li>python311Packages.powerline</li>
    <li>python311Packages.powerline.dist</li>
    <li>python311Packages.python-hglib</li>
    <li>python311Packages.python-hglib.dist</li>
    <li>python312Packages.glean-parser</li>
    <li>python312Packages.glean-parser.dist</li>
    <li>python312Packages.glean-sdk</li>
    <li>python312Packages.glean-sdk.dist</li>
    <li>python312Packages.python-hglib</li>
    <li>python312Packages.python-hglib.dist</li>
    <li>sourcehut.hgsrht</li>
    <li>sourcehut.hgsrht.dist</li>
  </ul>
</details>

**Comment by emilazy** [comment]
I assume the `lmdb` library is vendored through some Rust `*-sys` crate now?

**Comment by pyrox0** [comment]
Yes, it uses the [lmdb-rkv-sys](https://crates.io/crates/lmdb-rkv-sys) crate

**Comment by emilazy** [comment]
Do you know if we can easily get it to use the system `lmdb`? We prefer to do that when possible.

**Comment by pyrox0** [comment]
I'm not sure, as I've never used the crate before, and I'm not familiar with rust development. I'll look into it, but no guarantees that I can get anything working lol

**Pull request merged**